Add code for projecting to the boundary when using either Gauss or GL points

nilsdeppe opened the pull request:
## Proposed changes

Add code for projecting to the boundary when using either Gauss or Gauss-Lobatto points

Note that I'm adding this function to the evolution namespace since I don't know if or how the elliptic code will want to handle Gauss points since the reason this works in the evolution code is because I'm restructuring all the time derivative calculations and boundary terms.

Depends on:
- [x] #2534 

Only the following commits are new:
- Add ProjectToBoundary for evolution DG

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add ProjectToBoundary for evolution DG

nilsdeppe commented:
Rebased on develop. No longer dependent

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -0,0 +1,190 @@
+
+template <typename... Tags>
+struct NumberOfIndependentComponents<tmpl::list<Tags...>> {
+  static constexpr size_t value = (... + Tags::type::size());
```
🤯

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -0,0 +1,190 @@
+}  // namespace detail
+
+template <typename VolumeVarsTagsList, typename FaceVarsTagsList, size_t Dim>
+void contiguously_project_to_boundary(
```
- Add documentation
- minor suggestion: change name to `project_contiguous_data_to_boundary`, which to me is clearer that the "contiguous" is describing the data, not the projection.

nilsdeppe replied:
Holy smokes, sorry for completely missing documentation all together...

fmahebert replied:
Haha no worries, I was at least able to (mostly) understand this one even without docs ;)

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -0,0 +1,190 @@
+  using volume_vars_face_subset_list = tmpl::back<front_face_vars_split>;
+  static_assert(
+      std::is_same_v<volume_vars_face_subset_list, VolumeVarsTagsList>,
+      "The VolumeVarsTagsList must be a subset of the FaceVarsTagsList.");
```
Make sure the docs explain this business of extra tags and the tag ordering

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -0,0 +1,190 @@
+
+    DataVector face_view{
+        first_face_field[0].data(),
+        first_face_field[0].size() * number_of_independent_components};
```
Isn't all this just `DataVector face_view{face_fields->data(), face_fields->size()};` ? Or do you have to do it this way because of tag ordering? A comment might be helpful

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -0,0 +1,190 @@
+                   // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+                   DataVector{const_cast<double*>(first_volume_field[0].data()),
+                              first_volume_field[0].size() *
+                                  number_of_independent_components},
```
same question

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -0,0 +1,190 @@
+                     Spectral::Quadrature::GaussLobatto) == volume_mesh,
+           "The mesh must be either a uniform Gauss or Gauss-Lobatto mesh, but "
+           "got "
+               << volume_mesh);
```
- [non-binding] Mild suggestion for making a `uniform_gauss_lobatto_mesh` as you did before and using that in the assert
- Perhaps the message could be tweaked to clarify the operation would work in principle, but is just not implemented at this time?

nilsdeppe replied:
- I inlined it here because it's only needed in debug builds
- I agree, let me know if my wording is better

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -0,0 +1,190 @@
+    const size_t volume_grid_points = volume_mesh.number_of_grid_points();
+
+    const double* vars_data = volume_fields.data();
+    double* interface_vars_data = get<first_volume_tag>(*face_fields)[0].data();
```
same question as above

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -0,0 +1,190 @@
+    double* interface_vars_data = get<first_volume_tag>(*face_fields)[0].data();
+
+    // Run the SliceIterator as the outer-most loop since incrementing the slice
+    // iterator is surprisingly expensive.
```
For my understanding:
- why not use the function `data_on_slice` here? this loop appears to be copy-pasted from there... please add a comment explaining why you are duplicating code if it is truly unavoidable.
- have you newly measured the slice iterator to be expensive, or are you documenting here the knowledge from your experiments ~2-3 years ago?

nilsdeppe replied:
- I'll add a comment: can't be done because `data_on_slice` assumes the face and volume tags are equal
- I have not, but the iterator hasn't changed since I optimised it and still concluded that it sucks :P

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -0,0 +1,190 @@
+
+template <typename TagsToProjectList, typename VolumeVarsTagsList,
+          typename FaceVarsTagsList, size_t Dim>
+void project_tensors_to_boundary(
```
Add documentation

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -0,0 +1,190 @@
+                     Spectral::Quadrature::GaussLobatto) == volume_mesh,
+           "The mesh must be either a uniform Gauss or Gauss-Lobatto mesh, but "
+           "got "
+               << volume_mesh);
```
Make consistent changes here as in contiguous case

fmahebert commented on tests/Unit/Evolution/DiscontinuousGalerkin/Test_ProjectToBoundary.cpp:
```diff
@@ -0,0 +1,126 @@
+      expected_face_values = polynomial_volume_data(
+          interface_logical_coordinates(face_mesh, direction), powers);
+    }
+    const Scalar<DataVector> expected_var1{face_mesh.number_of_grid_points(),
```
[minor] Might be slightly clearer to call this `zeros` since you are using it in comparisons with Var1 and Var3. I don't feel strongly, up to you if you want to change it.

nilsdeppe replied:
I thought about this and the reason (which can reasonably be questioned) is I find the code easier to read when we have checks `a == expected_a` instead of `a == zero`. What do you think?

fmahebert replied:
The counter argument is `b == expected_a` :P. I don't know of a great solution here, so oh well, I'm happy for you to leave as is.

nilsdeppe replied:
haha self-rekt... Let me add a `expected_var3` :P

fmahebert reviewed (commented):
Overall the code looks good to me, I'd just like to see documentation added

nilsdeppe commented:
Rebased on develop and added a fixup commit. Thanks for the review!

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -68,6 +86,10 @@ void contiguously_project_to_boundary(
     auto& first_volume_field = get<first_volume_tag>(volume_fields);
 
+    // The size is the number of tensor components we are projecting times the
+    // number of grid points on the face. Note that this is _note_ equal to the
```
typo: note -> not

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -110,6 +140,13 @@ void contiguously_project_to_boundary(
 }
 
+/*!
+ * \brief Projects a subset of the tensor in the `volume_fields` onto the face
```
typo: tensor -> tensors

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -110,6 +140,13 @@ void contiguously_project_to_boundary(
+/*!
+ * \brief Projects a subset of the tensor in the `volume_fields` onto the face
+ *
+ * The tensor to project are listed in the `TagsToProjectList`.
```
typo: tensor -> tensors

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -0,0 +1,227 @@
+    const Mesh<Dim>& volume_mesh, const Direction<Dim>& direction) noexcept {
+  static_assert(tmpl::size<VolumeVarsTagsList>::value != 0,
+                "Must have non-zero number of volume fields");
+  static_assert(tmpl::size<FaceVarsTagsList>::value != 0,
```
- Instead of checking `!= 0`, could you check `>= tmpl::size<VolumeVarsTagsList>::value`? This would fail faster for lists of face tags that are obviously not supersets of the volume tags...
- In the same line of reasoning... would it make sense to add static_asserts that `tmpl::back<VolumeVarsTagsList>` and `tmpl::front<VolumeVarsTagsList>` are both in `FaceVarsTagsList`? Maybe this is getting to overkill.

fmahebert commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -0,0 +1,227 @@
+  static_assert(tmpl::size<VolumeVarsTagsList>::value != 0,
+                "Must have non-zero number of volume fields");
+  static_assert(tmpl::size<FaceVarsTagsList>::value != 0,
+                "Must have non-zero number of face fields");
```
Consider adding a static assert that each `TagsToProjectList` is in both other lists

nilsdeppe commented:
Rebased on develop to resolve conflicts and pushed fixup

fmahebert reviewed (commented):
Oops, I didn't mean to submit those previous comments until I had these ready to join them... sorry about giving you work twice.

I'm happy with the docs and changes so far. Here's a few opportunities to tighten up the static asserts.

I'm happy for you to squash :)

nilsdeppe commented:
Okay, added the helpful assertions and squashed. Thanks for all the great feedback :D

kidder commented on src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp:
```diff
@@ -0,0 +1,246 @@
+ * a boundary `Variables`
+ *
+ * The `volume_fields` are all projected into the `face_fields` in the direction
+ * `direction`. The variables in the volume (`VolumeVarsTagsList`) must be a
```
"The variables ... entire list must be a subset."  -> "The tags in `VolumeVarsTagsList` must be a contiguous subset of the tags in `FaceVarsTagsList`."

If you don't think that is clear enough, then add the sentence "That is, `FaceVarsTagsList` must be equivalent to `tmpl::append<Before, VolumeVarsTagsList, After>` where `Before` and `After` are `tmpl::list` of arbitrary size."

kidder reviewed (changes requested):
You can squash

nilsdeppe commented:
Rebased and squashed in the suggested rewording. I decided to include the extra verbose sentence because I'd rather have someone be "too clear" on something than confused :) Thanks for all the great feedback!
